sympy/functions/elementary/piecewise.py

```diff
@@ -1189,9 +1158,58 @@ def _clip(A, B, k):
 
 
 def piecewise_simplify_arguments(expr, **kwargs):
-    from sympy import simplify
+    from sympy.functions.elementary.complexes import im, re
+    from sympy import simplify, Interval, oo
+    # make conditions canonical
     args = []
     for e, c in expr.args:
+        # make c canonical
+        if (not c.is_Atom and not isinstance(c, Relational) and
+            not c.has(im, re)):
+            free = c.free_symbols
+            if len(free) == 1:
+                x = free.pop()
+                try:
+                    s = c.as_set()
+                    if s is S.Reals or isinstance(s, Interval) and (
+                            s.inf is -oo or s.sup is oo and
+                            c.has(oo) or c.has(-oo)):
+                        if s.inf is -oo and s.sup is oo:
+                            if c.xreplace({x: oo}) == True:
+                                if c.xreplace({x: s.inf}) == True:
+                                    c = x <= s.sup
+                                else:
+                                    c = x > s.inf
+                            elif c.xreplace({x: -oo}) == True:
+                                c = x < s.sup
+                            else:
+                                c = And(x > -oo, x < oo)
+                        elif s.inf is -oo:
+                            if c.xreplace({x: -oo}) == True:
+                                if c.xreplace({x: s.sup}) == True:
+                                    c = x <= s.sup
+                                else:
+                                    c = x < s.sup
+                            else:
+                                if c.xreplace({x: s.sup}) == True:
+                                    c = And(x > -oo, x <= s.sup)
+                                else:
+                                    c = And(x > -oo, x < s.sup)
+                        else:  # s.sup == oo
+                            if c.xreplace({x: oo}) == True:
+                                if c.xreplace({x: s.inf}) == True:
+                                    c = x >= s.inf
+                                else:
+                                    c = x > s.inf
+                            elif c.xreplace({x: s.inf}) == True:
+                                c = And(x >= s.inf, x < s.sup)
+                            else:
+                                c = And(x > s.inf, x < s.sup)
+                    else:
+                        c = s.as_relational(x)
+                except NotImplementedError:
```

Comment:
The try/except should be around precisely the statement from which the error is to be caught i.e. the `as_set` call.